src/lib/ppx_coda/versioned_module.ml

```diff
@@ -383,17 +371,180 @@ let version_module ~loc ~path:_ modname modbody =
+            in
+            (s, PStr [item'])
+        | [_] ->
+            (* should be unreachable *)
+            Location.raise_errorf ~loc
+              "Unrecognized pattern in deriving attribute"
+        | _ :: _ ->
+            (* should be unreachable *)
+            Location.raise_errorf ~loc "Duplicate deriving attribute"
+      in
+      let ptype_attributes' = deriving :: other_attrs in
+      let psig_desc =
+        Psig_type (recflag, [{type_ with ptype_attributes= ptype_attributes'}])
+      in
+      {sigitem with psig_desc}
+  | _ ->
+      sigitem
+
+let convert_module_type_signature signature =
+  List.map signature ~f:convert_module_type_signature_item
+
+let convert_module_type (mod_ty : module_type) =
+  match mod_ty.pmty_desc with
+  | Pmty_signature signature ->
+      let signature' = convert_module_type_signature signature in
+      {mod_ty with pmty_desc= Pmty_signature signature'}
+  | _ ->
+      Location.raise_errorf ~loc:mod_ty.pmty_loc
+        "Expected versioned module type to be a signature"
+
+type accum = {latest: string option; last: int option; sigitems: signature}
+
+let convert_module_decls ~loc:_ signature =
+  let init = {latest= None; last= None; sigitems= []} in
+  let f {latest; last; sigitems} sigitem =
+    match sigitem.psig_desc with
+    | Psig_module ({pmd_name; pmd_type; _} as pmd) ->
+        validate_module_version pmd_name.txt pmd_name.loc ;
+        let version = version_of_versioned_module_name pmd_name.txt in
+        Option.iter last ~f:(fun n ->
+            if Int.equal version n then
+              Location.raise_errorf ~loc:pmd_name.loc
+                "Duplicate versions in versioned modules" ;
+            if Int.( > ) version n then
+              Location.raise_errorf ~loc:pmd_name.loc
+                "Versioned modules must be listed in decreasing order" ) ;
+        let latest =
+          if Option.is_none latest then Some pmd_name.txt else latest
+        in
+        let psig_desc' =
+          Psig_module {pmd with pmd_type= convert_module_type pmd_type}
+        in
+        let sigitem' = {sigitem with psig_desc= psig_desc'} in
+        {latest; last= Some version; sigitems= sigitem' :: sigitems}
+    | _ ->
+        Location.raise_errorf ~loc:sigitem.psig_loc
+          "Expected versioned module declaration"
+  in
+  List.fold signature ~init ~f
+
+let version_module_decl ~loc ~path:_ modname signature =
+  Printexc.record_backtrace true ;
+  try
+    let open Ast_helper in
+    let modname = map_loc ~f:(check_modname ~loc:modname.loc) modname in
+    let {txt= {latest; sigitems; _}; _} =
+      map_loc ~f:(convert_module_decls ~loc:signature.loc) signature
+    in
+    let mk_module_decl name ty_desc =
+      Sig.mk ~loc (Psig_module (Md.mk ~loc name (Mty.mk ~loc ty_desc)))
+    in
+    let signature =
+      match latest with
+      | None ->
+          sigitems
+      | Some vn ->
+          let module E = Ppxlib.Ast_builder.Make (struct
```

Comment:
This looks like it may be unnecessary: it seems like `mk_module_decl` above is doing all the AST-building

Reply:
New code to generate `deserialize_binary_opt` and `versions` signature items needs it, I think.